lib/spack/spack/compilers/msvc.py

```
@@ -29,6 +29,78 @@
+        self._cmds = cmds
+
+    def __call__(self):
+        if not self._cmds:
+            raise RuntimeError(
+                """Attempting to run commands from CMD without specifying commands.
+                Please add commands to be run."""
+            )
+        out = subprocess.check_output(self.cmd_line, stderr=subprocess.STDOUT)  # novermin
+        return out.decode("utf-16le", errors="replace")  # novermin
+
+    @property
+    def cmd_line(self):
+        base_call = "cmd /u /c "
+        commands = " && ".join([str(x) for x in self._cmds])
+        # If multiple commands are being invoked by a single subshell
+        # they must be encapsulated by a double quote. Always double
+        # quote to be sure of proper handling
+        # cmd will properly resolve nested double quotes as needed
+        #
+        # `set`` writes out the active env to the subshell stdout,
+        # and in this context we are always trying to obtain env
+        # state so it should always be appended
+        return base_call + f'"{commands} && set"'
+
+    def add_command(self, cmd):
+        self._cmds.append(cmd)
+
+
+class VarsInvocation:
+    def __init__(self, script):
+        self._script = script
+
+    def __str__(self):
+        return f'"{self._script}"'
+
+    @property
+    def script(self):
+        return self._script
+
+
+class VCVarsInvocation(VarsInvocation):
+    def __init__(self, script, arch, msvc_version):
+        super(VCVarsInvocation, self).__init__(script)
+        self._arch = arch
+        self._msvc_version = msvc_version
+
+    @property
+    def sdk_ver(self):
+        if getattr(self, "_sdk_ver", None):
+            return self._sdk_ver + ".0"
+        return ""
+
+    @sdk_ver.setter
+    def sdk_ver(self, val):
+        self._sdk_ver = val
+
+    @property
+    def arch(self):
+        return self._arch
+
+    @property
+    def vcvars_ver(self):
+        return f"-vcvars_ver={self._msvc_version}"
+
+    def __str__(self):
```

IIRC `subprocess` is used vs. `Executable` because `cmd` is not a binary but rather a shell function: could you add an explanation for that.

Finally, the more I read, the more it seems like this object just exists to collect all the relevant bits of data needed to output a single formatted command invocation, so I think just replacing this with a function would be more straightforward.

If the object stays: `str` is typically for display purposes (not for creating anything structured that is meant to conform to an API) so I think this method should be renamed to something like `command_str`.

---

`cmd` is the shell itself. I'm actually not sure why this was used in this way, it was done before I was on the project in the first place (so like 2.5 years ago) and I just copied what was already there.

The reason I made this an object an opposed to making it a standalone function is because I needed a list of "invocable scripts" somewhere, and I needed a way to differentiate between them w.r.t. to MSVC and One API, and this solution was cleaner/more readable than a number of if statements, string parsing, and function calls. Functionally there's no real argument for object over function, just readability in my opinion. Also IMO I prefer the OOP interface style.

`__str__` was just a "consistent interface" function I could invoke, but you're very right, I'm kinda eschewing Python conventions, I'll rename as suggested.

---

Now that I look closer, I see that the `__init__` creates `CmdCall(*env_cmds)`. Is that invalid until we do `vcvars_call.sdk_ver = ...`? If so, then I think that the object should raise an exception when it comes time for `VCVarsInvocation` to retrieve `self.sdk_ver`. If not, then I think I see why the tension between `__init__`-time data and data that is only available later would encourage the use of an object, and I think documenting the cases where `sdk_ver` would reasonably be unset would be useful.

---

The `VCVarsInvocation` object does not require an SDK version (it's a mirror of a subset of the `vcvarsall` CLI so any optional arguments to that script are optional here), and a call without that argument is perfectly valid. In general, invoking the `vcvarsall` script associated with the version of MSVC this complier class is instantiated for, is sufficient to load the correct SDK and we don't need to provide a version to the call to `vcvarsall`, _except_ in the case where the SDK is part of the Spack DAG. In that case, a specific version of the SDK has either been specified directly or concretized based on some package requirement (or it's just the most recent, in which case we're doing the same thing as the script would be anyway) and to ensure we're building a compilation env with the correct SDK version, we must specify that argument. However we can't know that at construction time for the `Msvc` class. 

I'll make a comment summarizing the above a bit more concisely.

---

I believe this has been addressed.